lib/versioning/poetry/index.ts

```diff
@@ -10,6 +12,123 @@ export const urls = ['https://python-poetry.org/docs/versions/'];
 export const supportsRanges = true;
 export const supportedRangeStrategies = ['bump', 'extend', 'pin', 'replace'];
 
+// regex used by poetry.core.version.Version to parse union of SemVer
+// (with a subset of pre/post/dev tags) and PEP440
+// see: https://github.com/python-poetry/poetry-core/blob/01c0472d9cef3e1a4958364122dd10358a9bd719/poetry/core/version/version.py
```

Comment:
MIT License, so ok to copy
https://github.com/python-poetry/poetry-core/blob/master/LICENSE

split this file like in #10930, regex to `./patterns.ts` and helper functions to `./transform.ts`